streamflow/cwl/transformer.py

```diff
@@ -34,6 +35,89 @@ async def transform(
         return {self.get_output_name(): self._transform(*next(iter(inputs.items())))}
 
 
+class CartesianProductSizeTransformer(ManyToOneTransformer):
+    async def transform(
+        self, inputs: MutableMapping[str, Token]
+    ) -> MutableMapping[str, Token]:
+        for port_name, token in inputs.items():
+            if not isinstance(token.value, int) or token.value < 0:
+                raise WorkflowExecutionException(
+                    f"Step {self.name} got {token.value} on port {port_name}, but it must be a positive integer"
+                )
+        tag = get_tag(inputs.values())
+        value = functools.reduce(
+            lambda x, y: x * y, (token.value for token in inputs.values())
+        )
+        return {self.get_output_name(): Token(value, tag=tag)}
+
+
+class CloneListTokenTransformer(ManyToOneTransformer):
+    def __init__(self, name: str, workflow: Workflow, size_port: Port):
+        super().__init__(name, workflow)
+        self.add_input_port("__size__", size_port)
+
+    def _get_input_port_name(self) -> str:
+        return next(n for n in self.input_ports if n != "__size__")
+
+    def add_input_port(self, name: str, port: Port) -> None:
+        if len(self.input_ports) < 2 or name in self.input_ports:
+            super().add_input_port(name, port)
+        else:
+            raise WorkflowDefinitionException(
+                f"Step {self.name} must contain a single input port"
+            )
+
+    def get_input_port(self, name: str | None = None) -> Port:
+        return super().get_input_port(
+            self._get_input_port_name() if name is None else name
+        )
+
+    def get_size_port(self):
+        return self.get_input_port("__size__")
+
+    async def transform(
+        self, inputs: MutableMapping[str, Token]
+    ) -> MutableMapping[str, Token]:
+        # inputs has only two keys: __size__ and a port_name
+        if (
+            not isinstance(inputs["__size__"].value, int)
+            or inputs["__size__"].value < 0
+        ):
+            raise WorkflowExecutionException(
+                f"Step {self.name} got {inputs['__size__'].value} in the size port, but it must be a positive integer"
+            )
+        list_token = inputs[self._get_input_port_name()]
+        if not isinstance(list_token, ListToken):
+            raise WorkflowExecutionException(
+                f"Step {self.name} can clone only ListToken value, instead got a {type(list_token)}"
+            )
+        return {
+            self.get_output_name(): ListToken(
```

Comment:
This part is quite strange:
- Why you clone only `ListToken` objects?
- Why you retag the List elements but not the `ListToken` itself?
- This is not actually cloning a `Token` instance, but it is creating a list of cloned tokens, which is something very strange